blog/12-25-22-best-practices-for-writing-smart-contracts-with-real-world-examples/best-practices-for-writing-smart-contracts-with-real-world-examples.md

```diff
@@ -0,0 +1,534 @@
+    url: https://github.com/bobeu
+    image_url: https://github.com/bobeu.png
+tags: [solidity, celo, alfajores, smart-contract]
+hide_table_of_contents: true
+slug: /tutorials/best-practices-for-writing-smart-contracts-with-real-world-examples
+---
+
+![header](images/3.png)
+
+# Introduction
+
+The backbone of most web3 projects, if not all, is a well thought-out smart contract architecture that handles the sensitive aspect of the application such as financial transactions, health record of patients in a hospital, game logic and so on. Being a Celo developer, it is highly necessary to have a good engineering mindset when writing smart contracts to avoid costly mistakes that can put you at the mercy of the bad actors.
+
+# Prerequisites​
+
+We are going to examine some of the best practices, approaches and good standards for writing smart contracts that are targeted at EVM-compatible networks such as Celo. This tutorial is best suited for anyone with:
+-	At least foundational knowledge of smart contract development.
+-	Experience with other programming languages like C++ or Python will be an added advantage.
+
+# Requirements​
+
+We will leverage the following technologies and tools. Be sure to follow the steps accordingly.
+- VSCode
+- Nodejs >=v14.0.0
+- Hadhat.
+
+> Note: We will not cover testing in this tutorial. For testing, please refer to [this tutorial](https://docs.celo.org/blog/tutorials/how-to-write-unit-testing-for-contracts-with-hardhat)
+
+**Content**
+
+Some of the smart contract best practices we will discuss are as follows:
+
+- Avoid floating pragma.
+- Modular approach.
+- Circuit breaker.
+- External contract interaction.
+- Upgradability.
+- Check, effect, interaction.
+- Adopt libraries where possible.
+- Use Speed breaks where necessary.
+- Proper use of assert(), require() and revert() functions.
+- Use modifiers properly.
+- Integer division.
+- Fallback/receive function.
+- Explicit labeling of function visibility.
+- Use interface type for accepting contract addresses.
+- Use standard interface (S).
+
+**General consideration**
+
+In programming, it is assumed there is no software that is perfect. As a web3 developer, you should try to be ahead of the bad players. That is, finding, identifying and fixing loopholes in your code before others find it. Without further ado, let’s get started.
+
+We will write fresh match-making contracts that connect and manage the processes of marriage between two opposite sexes. It features the following subjects:
+
+- HusbandToBe:  Will have the following properties.
+	- Inlaw.
+	- Wife
+	- Profile: struct - 
+		```bash
+				age
+				isMale
+				Bank - e.g Cash in bank balance etc
+				Property : Owned tangible assets. 
+				Religion: What they believe in.
+				Nature: Whether they are drunkard or otherwise.
+				Status: Type “Enum” showing their status as single or taken or married.
+		```
+- WifeToBe: This contract will have similar properties such as: 
```

Comment:
I think there is an intendation issue here. Please look into it.

Reply:
Yeah, Corrected.